### ISSUE TITLE
package requires Go 1.2

### ISSUE THREAD

**Issue by stapelberg** [sorcix/irc#4]
Because `strings.IndexByte` is used, Go 1.2 is required to use sorcix/irc. This is a bit unfortunate, as Debian stable only ships Go 1.0.

Are you dead set on using `strings.IndexByte` or would you be willing to accept a PR (or change the code yourself if you prefer) which makes sure only functions from Go 1 are used?

**Comment by sorcix** [comment]
IndexByte is implemented in asm for speed. I want to keep it fast, as handling IRC messages is something that happens very often. Let's do a benchmark and check the speed difference.

**Comment by Merovius** [comment]
Could we use build-tags to distinguish the cases? i.e. build a wrapper with two implementations, on go >= 1.2 it just calls strings.IndexByte and on < 1.2 it uses a custom go version?

**Comment by Merovius** [comment]
An alternative could be, to just copy the relevant asm-routines out of the runtime source and slap a // build +go1.2 on it, I *think* that should even give a *fast* IndexByte on older go versions (though I have only a limited understanding of how the go toolchain *actually* handles asm…).

**Comment by sorcix** [comment]
It should be possible to create the opposite of +go1.2 using a combination of build tags.

I'm looking for examples on how other projects solve this.

**Comment by danryan** [comment]
You can negate build tags by prefixing with `!`. For instance, to support go1.2 and up, move the specific code to a file, and add this to the top:

```
// +build go1.2

package irc
// ...
```

Then in another file, write the code that should only build with versions lower than go1.2:

```
// +build !go1.2

package irc
// ...
```

I tested this with go versions 1.4 and 1.1, albeit a very simple one :) Hope this helps!

**Comment by sorcix** [comment]
For the record, [I have asked for suggestions on reddit](https://www.reddit.com/r/golang/comments/2tcp78/question_whats_the_idiomatic_way_to_support_older/). I agree with the current comments saying that you really shouldn't use Go 1.0. That said, there are a lot of debian users and this change is really small.

Benchmarks before change:
```
$ go test -bench="."
PASS
BenchmarkPrefix_String_short	200000000	         7.75 ns/op	       0 B/op	       0 allocs/op
BenchmarkPrefix_String_long	 2000000	       745 ns/op	      64 B/op	       2 allocs/op
BenchmarkParsePrefix_short	 3000000	       446 ns/op	      48 B/op	       1 allocs/op
BenchmarkParsePrefix_long	 3000000	       498 ns/op	      48 B/op	       1 allocs/op
BenchmarkMessage_String	 1000000	      1698 ns/op	     176 B/op	       2 allocs/op
BenchmarkParseMessage_short	 1000000	      1595 ns/op	     128 B/op	       4 allocs/op
BenchmarkParseMessage_medium	  500000	      2450 ns/op	     192 B/op	       5 allocs/op
BenchmarkParseMessage_long	  500000	      3050 ns/op	     272 B/op	       5 allocs/op
ok  	_/home/vic/Repositories/irc	14.524s
```

Benchmarks after change, wrapped strings.IndexByte for >=1.2:
```
$ go test -bench="."
DEBUG: init() function in strings.go
PASS
BenchmarkPrefix_String_short	200000000	         7.74 ns/op	       0 B/op	       0 allocs/op
BenchmarkPrefix_String_long	 2000000	       740 ns/op	      64 B/op	       2 allocs/op
BenchmarkParsePrefix_short	 3000000	       464 ns/op	      48 B/op	       1 allocs/op
BenchmarkParsePrefix_long	 3000000	       498 ns/op	      48 B/op	       1 allocs/op
BenchmarkMessage_String	 1000000	      1664 ns/op	     176 B/op	       2 allocs/op
BenchmarkParseMessage_short	 1000000	      1637 ns/op	     128 B/op	       4 allocs/op
BenchmarkParseMessage_medium	  500000	      2529 ns/op	     192 B/op	       5 allocs/op
BenchmarkParseMessage_long	  500000	      3113 ns/op	     272 B/op	       5 allocs/op
ok  	_/home/vic/Repositories/irc	14.655s
```

Benchmark after change, replacement for strings.IndexByte for < 1.2:
```
$ go test -bench="."
DEBUG: init() function in strings_legacy.go
PASS
BenchmarkPrefix_String_short	200000000	         7.75 ns/op	       0 B/op	       0 allocs/op
BenchmarkPrefix_String_long	 2000000	       795 ns/op	      64 B/op	       2 allocs/op
BenchmarkParsePrefix_short	 2000000	       625 ns/op	      48 B/op	       1 allocs/op
BenchmarkParsePrefix_long	 2000000	       723 ns/op	      48 B/op	       1 allocs/op
BenchmarkMessage_String	 1000000	      1737 ns/op	     176 B/op	       2 allocs/op
BenchmarkParseMessage_short	 1000000	      1754 ns/op	     128 B/op	       4 allocs/op
BenchmarkParseMessage_medium	  500000	      2854 ns/op	     192 B/op	       5 allocs/op
BenchmarkParseMessage_long	  500000	      3896 ns/op	     272 B/op	       5 allocs/op
ok  	_/home/vic/Repositories/irc	15.785s
```

Looks like there isn't really a speed difference for users on the latest version. The weird thing is that `go build -gcflags -m` does not mention being able to inline my indexByte function.

**Comment by sorcix** [comment]
Fix in be4b6a03b1e467837f5224715ec3a97ac7b4e0fd. Any comments?

**Comment by stapelberg** [comment]
The fix looks good to me :). Thanks a lot for doing this!
